### PR TITLE
Product variants OLD PR

### PR DESCRIPTION
**Pull request by bricht**
-Added string type or message entity id to accommodate for product variants

### Description (*)
PR moved to https://github.com/magento/catalog-storefront/pull/413

### Code Review Checklist (*)

See dataied [checklist](https://github.com/magento/catalog-storefront/wiki/Code-Review-checklist)

- [ ] Story AC is completed
- [ ] proposed changes correspond to [Magento Technical Vision](https://devdocs.magento.com/guides/v2.2/coding-standards/technical-guidelines.html)
- [ ] new or changed code is covered with web-api/integration tests (if applicable)
  - expected results in test verified with data from fixture
- [ ] no backward incompatibile changes
- [ ] Export API (et_schema.xml) and SF API schemas (proto schema) are reflected in the codebase
  - prerequisite: story branch created with all needed generated classes according to proposes schema-changes
  - DTO classes do not contain any manual changes (Magento\CatalogExportApi\*, Magento\CatalogStorefrontApi\*)
- [ ] Class usage: magento/catalog-storefront repo don't use directly classes from magento/saas-export repo and vise-verse
  - Check composer.json dependencies
- [ ] Legacy code is deleted
  - Any Data Providers present in Connector part  (Magento\CatalogStorefrontConnector, Magento\*Extractor modules)
  - And Data Providers from Export API (magento/saas-export repo) that is not relevant anymore
  - Any DTO for Export API/SF API which does not reflect current schema: et_schema, proto schema
  - Any “mapper” on Message Broker (between Export API and SF API)
    - if mapper still needed, verify fields used in mapping, remove not relevant fields

### CODE REVIEW

File: app/code/Magento/CatalogStorefront/Model/CatalogRepository.php [hunk 1]
```diff
@@ -67,13 +68,15 @@ public function __construct(
...
      * @throws CouldNotSaveException
+     * @throws \RuntimeException
      */
     public function saveToStorage(array $dataPerType): void
     {
         foreach ($dataPerType as $entityType => $dataPerStore) {
             foreach ($dataPerStore as $storeCode => $data) {
                 $sourceName = $this->storageState->getCurrentDataSourceName([$storeCode, $entityType]);
                 $this->deleteEntities($data[self::DELETE] ?? [], $sourceName, $entityType);
+                $this->deleteEntitiesByQuery($data[self::DELETE_BY_QUERY] ?? [], $sourceName, $entityType);
```
[the remark this answers is not in the document]
Comment (bricht) [reply app/code/Magento/CatalogStorefront/Model/CatalogRepository.php:79]: They have different implementations - one does bulkDelete(entity_ids)
 and the other - deleteByQuery(queryParams).

File: app/code/Magento/CatalogMessageBroker/Model/FetchProductVariants.php [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+declare(strict_types=1);
+
+namespace Magento\CatalogMessageBroker\Model;
+
+use Magento\CatalogMessageBroker\HttpClient\RestClient;
+use Magento\CatalogExport\Event\Data\Entity;
+use Psr\Log\LoggerInterface;
+
+/**
+ * @inheritdoc
```
Comment (RuslanKostiv1) [inline app/code/Magento/CatalogMessageBroker/Model/FetchProductVariants.php:15]: @inheritdoc should be used only on class members

Comment (bricht) [reply app/code/Magento/CatalogMessageBroker/Model/FetchProductVariants.php:15]: Resolved

File: dev/tests/api-functional/testsuite/Magento/CatalogMessageBroker/Model/MessageBus/ProductVariantsTest.php [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+    private function getProduct(string $sku): ProductInterface
+    {
+        try {
+            return $this->productRepository->get($sku);
+        } catch (NoSuchEntityException $e) {
+            throw new NoSuchEntityException();
+        }
+    }
+
+    /**
```
[the remark this answers is not in the document]
Comment (bricht) [reply dev/tests/api-functional/testsuite/Magento/CatalogMessageBroker/Model/MessageBus/ProductVariantsTest.php:190]: Resolved

File: dev/tests/api-functional/testsuite/Magento/CatalogMessageBroker/Model/MessageBus/ProductVariantsTest.php [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+        // after the records have been deleted by the _delete_by_query call.
+        //See https://www.elastic.co/guide/en/elasticsearch/reference/6.8/docs-refresh.html#docs-refresh
+        sleep(3);
+
+        $this->expectException(\InvalidArgumentException::class);
+        $this->expectExceptionMessage(sprintf(self::ERROR_MESSAGE, $configurableId));
+        $this->variantService->GetProductVariants($this->variantsGetRequestInterface);
+    }
+
+    /**
```
[the remark this answers is not in the document]
Comment (bricht) [reply dev/tests/api-functional/testsuite/Magento/CatalogMessageBroker/Model/MessageBus/ProductVariantsTest.php:154]: Resolved

File: dev/tests/api-functional/testsuite/Magento/CatalogMessageBroker/Model/MessageBus/ProductVariantsTest.php [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+                $feedItem['product_id'] = $childId;
+                unset($feedItem['modifiedAt'], $feedItem['deleted'], $feedItem['parent_id']);
+                return $feedItem;
+            }, $feedData);
+        } catch (\Exception $e) {
+            $this->fail('Feed data did not match the expected format');
+        }
+    }
+
+    /**
```
[the remark this answers is not in the document]
Comment (bricht) [reply dev/tests/api-functional/testsuite/Magento/CatalogMessageBroker/Model/MessageBus/ProductVariantsTest.php:174]: Resolved

File: dev/tests/api-functional/testsuite/Magento/CatalogMessageBroker/Model/MessageBus/ProductVariantsTest.php [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+            ProductVariantRequestInterface::class
+        );
+        $this->messageBuilder = Bootstrap::getObjectManager()->create(ChangedEntitiesMessageBuilder::class);
+        $this->productVariantFeed = Bootstrap::getObjectManager()->get(FeedPool::class)->getFeed('variants');
+        $this->productRepository = Bootstrap::getObjectManager()->create(ProductRepositoryInterface::class);
+        $this->responseArrayMapper = Bootstrap::getObjectManager()->create(ProductVariantResponseArrayMapper::class);
+    }
+
+    /**
+     * Validate something
```
[the remark this answers is not in the document]
Comment (bricht) [reply dev/tests/api-functional/testsuite/Magento/CatalogMessageBroker/Model/MessageBus/ProductVariantsTest.php]: Resolved

File: app/code/Magento/CatalogStorefront/Model/VariantService.php [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+    {
+        $deleteFields = \array_map(function ($id) {
+            return ['id' => $id];
+        }, $request->getId());
+
+        $variantsInElasticFormat = [
+            'product_variant' => [
+                //TODO: Adapt to work without store code
+                self::EMPTY_STORE_CODE => [
+                    'delete_by_query' => $deleteFields
```
[the remark this answers is not in the document]
Comment (bricht) [reply app/code/Magento/CatalogStorefront/Model/VariantService.php]: Resolved

File: app/code/Magento/CatalogMessageBroker/Model/MessageBus/Product/DeleteProductsConsumer.php [hunk 8]
```diff
@@ -49,7 +49,7 @@ public function __construct(
     /**
      * @inheritdoc
      */
-    public function execute(array $entities, string $scope): void
+    public function execute(array $entities, string $scope = null): void
```
[the remark this answers is not in the document]
Comment (bricht) [reply app/code/Magento/CatalogMessageBroker/Model/MessageBus/Product/DeleteProductsConsumer.php]: Resolved

File: app/code/Magento/CatalogMessageBroker/Model/MessageBus/ConsumerEventInterface.php [hunk 9]
```diff
@@ -18,9 +18,9 @@ interface ConsumerEventInterface
      * Execute consumers by ids for specified scope
      *
      * @param Entity[] $entities
-     * @param string $scope
+     * @param string|null $scope
      *
      * @return void
      */
-    public function execute(array $entities, string $scope): void;
+    public function execute(array $entities, string $scope = null): void;
```
[the remark this answers is not in the document]
Comment (bricht) [reply app/code/Magento/CatalogMessageBroker/Model/MessageBus/ConsumerEventInterface.php]: Resolved

File: app/code/Magento/CatalogMessageBroker/Model/MessageBus/Category/PublishCategoriesConsumer.php [hunk 10]
```diff
@@ -78,7 +78,7 @@ public function __construct(
     /**
      * @inheritdoc
      */
-    public function execute(array $entities, string $scope): void
+    public function execute(array $entities, string $scope = null): void
```
[the remark this answers is not in the document]
Comment (bricht) [reply app/code/Magento/CatalogMessageBroker/Model/MessageBus/Category/PublishCategoriesConsumer.php]: Resolved

File: app/code/Magento/CatalogMessageBroker/Model/MessageBus/Category/DeleteCategoriesConsumer.php [hunk 11]
```diff
@@ -49,7 +49,7 @@ public function __construct(
     /**
      * @inheritdoc
      */
-    public function execute(array $entities, string $scope): void
+    public function execute(array $entities, string $scope = null): void
```
[the remark this answers is not in the document]
Comment (bricht) [reply app/code/Magento/CatalogMessageBroker/Model/MessageBus/Category/DeleteCategoriesConsumer.php]: Resolved

File: app/code/Magento/CatalogExport/Event/Data/Entity.php [hunk 12]
```diff
@@ -14,7 +14,7 @@
 class Entity
 {
     /**
-     * @var int
+     * @var int|string
      */
```
[the remark this answers is not in the document]
Comment (bricht) [reply app/code/Magento/CatalogExport/Event/Data/Entity.php:18]: Resolved

File: app/code/Magento/CatalogMessageBroker/Model/MessageBus/Product/PublishProductsConsumer.php [hunk 13]
```diff
@@ -73,7 +73,7 @@ public function __construct(
     /**
      * @inheritdoc
      */
-    public function execute(array $entities, string $scope): void
+    public function execute(array $entities, string $scope = null): void
```
[the remark this answers is not in the document]
Comment (bricht) [reply app/code/Magento/CatalogMessageBroker/Model/MessageBus/Product/PublishProductsConsumer.php]: Resolved

File: app/code/Magento/CatalogStorefront/Model/VariantService.php [hunk 14]
```diff
@@ -0,0 +1,279 @@
...
+        try {
+            $variantsInElasticFormat = [];
+            foreach ($request->getVariants() as $variantData) {
+                $optionValues = $variantData->getOptionValues();
+                $id = $variantData->getId();
+                $explodedId = explode('/', $id);
+                $parentId = $explodedId[1];
+                $childId = $explodedId[2];
+                foreach ($optionValues as $optionValue) {
+                    preg_match('/(?<=\:)(.*)(?=\/)/', $optionValue, $match);
```
[the remark this answers is not in the document]
Comment (bricht) [reply app/code/Magento/CatalogStorefront/Model/VariantService.php]: Resolved

File: app/code/Magento/CatalogStorefront/Model/Storage/Client/QueryInterface.php [hunk 15]
```diff
@@ -57,4 +57,22 @@ public function getEntries(
         array $ids,
         array $fields
     ): EntryIteratorInterface;
+
+    /**
+     * Search entries by specified search arguments.
+     * Search works using the "must match" logic.
```
[the remark this answers is not in the document]
Comment (bricht) [reply app/code/Magento/CatalogStorefront/Model/Storage/Client/QueryInterface.php:63]: Extended to support both filter and match.

File: app/code/Magento/CatalogStorefront/Model/Storage/Data/SearchResultIteratorFactory.php [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+    public function __construct(
+        ObjectManagerInterface $objectManager,
+        DocumentFactory $documentFactory
+    ) {
+        $this->objectManager = $objectManager;
+        $this->documentFactory = $documentFactory;
+    }
+
+    /**
+     * Create class instance with specified parameters
```
[the remark this answers is not in the document]
Comment (bricht) [reply app/code/Magento/CatalogStorefront/Model/Storage/Data/SearchResultIteratorFactory.php]: Resolved

**Pull request merged**